[SE-0351] Revise regex builder proposal (#1634)
apple/swift-evolution: proposals/0351-regex-builder.md

### `RegexComponent` protocol

> One of the goals of the regex builder DSL is allowing the developers to easily compose regexes from common currency types and literals, or even define custom patterns to use for matching. We introduce `RegexComponent` in the Swift Standard Library, a protocol that unifies all types that can represent a component of a regex. Since regexes are composable, the `Regex` type itself conforms to `RegexComponent`.

Reviewer: I don't know if we've formally laid out what is stdlib vs the `Swift` module, and whether the regex builders modules is considered stdlib (and if not, should be it be considered as?).

```suggestion
One of the goals of the regex builder DSL is allowing the developers to easily compose regexes from common currency types and literals, or even define custom patterns to use for matching. We introduce `RegexComponent` in the `Swift` module, a protocol that unifies all types that can represent a component of a regex. Since regexes are composable, the `Regex` type itself conforms to `RegexComponent`.
```

Reviewer: It might make sense to mention the new module first too, maybe a first section like:

### `RegexBuilder` module

...

Reviewer: FWIW, I say "stdlib" or "stdlib proper" as a stand-in for the `Swift` module; I'm much more guilty than most. I honestly don't know what terminology to use. Since this proposal is one of the first to need that distinction, let's be as clear as we can be.

Reviewer 2: I think that it will take a little time to iron out the nomenclature here, so let's go with "`Swift` module" for now (since we want to unblock this quickly); it is at least precise and unambiguous.

Author: FWIW, saying "the `Swift` module` means you would be able to do a qualified name lookup such as `Swift.Regex`, which doesn't work in our case. Nonetheless, I've used "the `Swift` module" for now.

```swift
public protocol RegexComponent<RegexOutput> {
  associatedtype RegexOutput
  var regex: Regex<Output> { get }
}

extension Regex: RegexComponent {
  public typealias RegexOutput = Output
  public var regex: Regex<Output> { self }
}
```

Note:

> - `RegexComponent` and `Regex`'s conformance to `RegexComponent` are available without importing `RegexBuilder`. All other types and conformances introduced in this proposal are in the `RegexBuilder` module.

Author: > Is this where we landed, with the `String` conformance not being in the stdlib proper (i.e. `Swift`)?

Yes.

> Either way, let's be very explicit what we mean by "other types".

"All other types" seems pretty clear to me. The sentence below it already says explicitly:

> By conforming standard library types to `RegexComponent`, we allow them to be used inside the regex builder DSL as a match target. **These conformances are available in the `RegexBuilder` module.**

- The associated type `RegexOutput` intentionally has a `Regex` prefix. `Output` would cause confusion in standard library conforming types such as `String`, i.e. `String.Output`.

> By conforming standard library types to `RegexComponent`, we allow them to be used inside the regex builder DSL as a match target. These conformances are available in the `RegexBuilder` module.

Reviewer: And also as arguments to the string processing algorithms (one reason to want the conformance in `Swift`).

Reviewer: We resolve this: I had missed the fact the collection generic versions would apply instead of the regex ones. So it makes sense to keep the conformances in the regex builder module.

} // Regex<(Substring, Int, Float, Substring)>
```

The following example creates a regex by concatenating subpatterns.

```swift
let regex = Regex {
  "regex"
  "builder"
  "is"
  "easy"
}
let match = try regex.prefixMatch(in: "regex builder is easy!")

> match?.0 // => "regex builder is easy"

Author: Fixed by adding whitespaces.

```

<details>

> <summary>API definition</summary>

Reviewer: I think, in general at least, we want to hoist the struct decls out of the details triangle. The stuff to go in the triangle would be the speculation about how variadic generics work (if it even belongs in this proposal) and the result builder overloads and machinery.

Author: All struct declarations already have an abbreviated version outside the collapsible. Inside the collapsible there's a full declaration along with initialize roverloads.

Reviewer: But why not just put the declaration outside? Stick the comments about what we would have done if variadic generics existed and the machinery inside the disclosure. I don't think we should be hiding end-developer API inside the triangles

`buildOptional` and `buildEither` are intentionally not supported due to ergonomic issues and fundamental semantic differences between regex conditionals and result builder conditionals. Please refer to the [alternatives considered](#support-buildoptional-and-buildeither) section for detailed rationale.

</details>

### Capture and reference

Capture is a common regex feature that saves a portion of the input upon match. In regex builder, `Capture` and `TryCapture` are regex components that produce a new regex by inserting the captured pattern's whole match (`.0`) to the `.1` position of `Output`. When a transform closure is provided, the whole match (`.0`) of the captured content will be transformed to using the closure.

```swift
public struct Capture<Output>: RegexComponent { ... }
public struct TryCapture<Output>: RegexComponent { ... }
```

To do a simple capture, you provide `Capture` with a regex component or a regex component builder closure.

```swift
// Equivalent: '(CREDIT|DEBIT)'
Capture {
  ChoiceOf {
    "CREDIT"
    "DEBIT"
  }
} // `.RegexOutput == (Substring, Substring)`
```

> To transform the captured substring, specify a `transform:` closure.

Reviewer: ```suggestion
A capture will be represented in the type signature as a slice of the input, i.e. `Substring`. To transform the captured substring into another value during matching, specify a `transform:` closure.
```

```
// This example is similar to the one above, however in this example we
// transform the result of the capture into:
// "Transaction Kind: CREDIT" or "Transaction Kind: DEBIT"
Capture {
  ChoiceOf {
    "CREDIT"
    "DEBIT"
  }
} transform: {
  "Transaction Kind: \($0)"
} // `.RegexOutput == (Substring, String)`
```

> The transform closure can throw. When a transform closure throws during matching, the matching will abort and the error will be propagated directly to the top-level matching API that's being called, e.g. `Regex.wholeMatch(in:)` and `Regex.prefixMatch(in:)`.

Reviewer: ```suggestion
The transform closure can throw. When a transform closure throws during matching, the matching will abort and the error will be propagated directly to the top-level matching API that's being called, e.g. `Regex.wholeMatch(in:)` and `Regex.prefixMatch(in:)`. Aborting is useful for cases where you know that matching can never succeed or when you detect that an important invariant has been violated and the matching procedure needs to be aborted.
```

> An alternative version of capture is called `TryCapture`, which works in cases where you want to transform the capture, but the transformation may return nil. When a nil is returned, the regex engine backtracks and tries an alternative. `TryCapture` is most useful when you transform a captured substring by calling a failable initializer.

Reviewer: ```suggestion
An alternative version of capture is called `TryCapture`, which works in cases where you want to transform the capture, but the transformation may return nil. When a nil is returned, the regex engine backtracks and tries an alternative. For example, `TryCapture` makes it easy to directly call a failable initializer during matching.
```

```swift
enum TransactionKind: String {
  case credit = "CREDIT"
  case debit = "DEBIT"
}

TryCapture {
  ChoiceOf {
    "CREDIT"
    "DEBIT"
  }
} transform: {
  // This initializer may return nil which is why we used TryCapture.
  TransactionKind(rawValue: String($0))
}
```

> 

Reviewer: Maybe make reference treatment a sub-section here? Or should we finish non-reference captures here and make this the start of the next section?

Author: I made reference a new section after `Capture`.

Reference is a feature that can be used to achieve named captures and named backreferences from textual regexes. Simply state what type the reference will hold on to and you can use it later once you've matched a string to get back a specific capture. Note the type you pass to reference will be whatever the result of a capture's transform is. A capture with no transform always has a reference type of `Substring`.

```swift
let kind = Reference(Substring.self)

let regex = Capture(as: kind) {
  ChoiceOf {
    "CREDIT"
    "DEBIT"
  }
}

let input = "CREDIT"
if let result = regex.firstMatch(in: input) {
  print(result[kind]) // CREDIT
}
```

> When capturing some regex with a reference specified, the reference will refer to the most recently captured content. The reference itself can also be used as a regex to match the most recently captured content, or as a name to look up the result of matching.

Reviewer: ```suggestion
Capturing stores the most recently captured content and references can be used as a name to look up the result of matching. The reference itself can also be used within a regex (commonly called a "backreference") to match the most recently captured content during matching.
```

```swift
let a = Reference(Substring.self)
let b = Reference(Substring.self)
let regex = Regex {
  Capture("abc", as: a)
  Capture("def", as: b)
  a
  Capture(b)
}

if let result = regex.firstMatch(in: input) {
  print(result[a]) // => "abc"
  print(result[b]) // => "def"
}
```

> A regex is considered invalid when it contains a use of reference without it ever being captured in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the use of a reference in a `Regex.Match.subscript(_:)` must have been captured in the regex that produced the match.

Reviewer: I think it's important to very clearly differentiate the `throw` of a runtime error during the match call and the trap that's generated in the subscript. As written, the reader has to read a little between the lines here.

Reviewer: Wait, is that what happens? I thought it would just give `nil`.

Reviewer: Can we clarify what's going on? I would find it incredibly developer hostile for this to trap if an alternation branch wasn't taken. Perhaps this should say "trap if the reference was not used in the regex" rather than captured.

Author: > Can we also be clear that this means that a reference that is captured only in an optional portion of a `Regex` can trap when accessed? Do we have API for knowing whether or not it is safe to access?

This behavior is unfortunate and could definitely use some more discussion.  I see some options:
1. Make the subscript return optional, but it could make things harder to use than numbered captures for non-optional captures.
    ```swift
    let ref = Reference(Substring.self)
    let regex = Regex {
          Capture(..., as: ref)
    }
    let match = try regex.fisrtMatch(of: ...)!
    match.1 // Substring
    match[ref] // Substring?
    ```
2. Provide an API to check whether a match result contains a reference. The user would be responsible for checking when the reference capture is in an optional branch of the regex. This can keep non-optional captures simple.
    ```swift
    // Non-optional trapping version
    subscript<Capture>(_ reference: Reference<Capture>) -> Capture { get }
    // Checking version
    subscript<Capture>(ifExists reference: Reference<Capture>) -> Capture? { get }
    // Predicate
    func contains<Capture>(_ reference: Reference<Capture>) -> Bool
    ```

    ```swift
    let ref = Reference(Substring.self)
    let regex = Regex {
          ZeroOrMore(Capture(..., as: ref))
    }
    let match = try regex.fisrtMatch(of: ...)!
    match.1 // Substring?
    match[ref] // Substring or trap
    match[ifExists: ref] // Substring?
    match.contains(ref) // Bool
    ```

Author: Went with option 1.

<details>
<summary>API definition</summary>

```swift
public struct Capture<Output>: RegexComponent {
  public var regex: Regex<Output> { get }
}

public struct TryCapture<Output>: RegexComponent {
  public var regex: Regex<Output> { get }
}
```

Below are `Capture` and `TryCapture` initializer variants on capture arity 0. Higher capture arities are omitted for simplicity.

>   

Reviewer: This would be where I'd start the disclosure triangle.

Author: This sentence is already in a collapsible.  The type declarations are both outside the collapsible at the top of the "Capture and reference" section (abbreviated), and within the collapsible (full).

Reviewer: The point was that the triangle would happen at this point, not earlier

  
  public init<R: RegexComponent, W>(
    as reference: Reference<W>,
    @RegexComponentBuilder _ component: () -> R
  ) where Output == (Substring, W), R.RegexOutput == W

  // ... `O(arity)` overloads
}
  
extension TryCapture {
  public init<R: RegexComponent, W, NewCapture>(
    _ component: R,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W
  
  public init<R: RegexComponent, W, NewCapture>(
    _ component: R,
    as reference: Reference<NewCapture>,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W
  
  public init<R: RegexComponent, W, NewCapture>(
    @RegexComponentBuilder _ component: () -> R,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W
  
  public init<R: RegexComponent, W, NewCapture>(
    as reference: Reference<NewCapture>,
    @RegexComponentBuilder _ component: () -> R,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W

  // ... `O(arity)` overloads
}

> /// A reference to a regex capture.

Reviewer: Yeah, I really feel like references deserve their own section.

```swift
let kind = Reference(Substring.self)

let regex = Capture(as: kind) {
  ChoiceOf {
    "CREDIT"
    "DEBIT"
  }
}

let input = "CREDIT"
if let result = regex.firstMatch(in: input) {
  print(result[kind]) // Optional("CREDIT")
}
```

Capturing stores the most recently captured content, and references can be used as a name to look up the result of matching. The reference itself can also be used within a regex (commonly called a "backreference") to match the most recently captured content during matching.

```swift
let a = Reference(Substring.self)
let b = Reference(Substring.self)
let c = Reference(Substring.self)
let regex = Regex {
  Capture("abc", as: a)
  Capture("def", as: b)
  ZeroOrMore {
    Capture("hij", as: c)
  }
  a
  Capture(b)
}

if let result = regex.firstMatch(in: "abcdefabcdef") {
  print(result[a]) // => Optional("abc")
  print(result[b]) // => Optional("def")
  print(result[c]) // => nil
}
```

> A regex is considered invalid when it contains a use of reference without it ever being captured as (used as `Capture(..., as: ...)` or `Capture(as: ...) { ... }`) in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the argument to a `Regex.Match.subscript(_:)` must have been captured as in the regex that produced the match.

Reviewer 3: I still think this language conflates two meanings of "captured". The meaning here is that it needs to be passed in a call to Capture or TryCapture, but the other way we use "captured" is that it matches some portion (even an empty range) of the input text. Can we make that more clear, here and in the subscript docs?

    _ component: R,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W
  
  public init<R: RegexComponent, W, NewCapture>(
    @RegexComponentBuilder _ component: () -> R,
    transform: @escaping (Substring) throws -> NewCapture?
  ) where Output == (Substring, NewCapture), R.RegexOutput == W

  // ... `O(arity)` overloads
}
```

</details>

### Mapping Output

In addition to transforming individual captures within a regex, you can also map the output of an entire regex to a different output type. You can use the `mapOutput(_:)` methods to reorder captures, flatten nested optionals, or create instances of a custom type.

This example shows how you can transform the output of a regex with three capture groups into an instance of a custom `SemanticVersion` type, matching strings such as `"1.0.0"` or `"1.0"`:

```swift
struct SemanticVersion: Equatable {
  var major, minor, patch: Int
}

let semverRegex = Regex {
  TryCapture(OneOrMore(.digit)) { Int($0) }
  "."
  TryCapture(OneOrMore(.digit)) { Int($0) }
  Optionally {
    "."
    TryCapture(OneOrMore(.digit)) { Int($0) }
  }

> }.map { _, c1, c2, c3 in

Reviewer: ```suggestion
}.mapOutput { _, c1, c2, c3 in
```

</details>

### Mapping Output

In addition to transforming individual captures within a regex, you can also map the output of an entire regex to a different output type. You can use the `mapOutput(_:)` methods to reorder captures, flatten nested optionals, or create instances of a custom type.

This example shows how you can transform the output of a regex with three capture groups into an instance of a custom `SemanticVersion` type, matching strings such as `"1.0.0"` or `"1.0"`:

```swift

> struct SemanticVersion: Equatable {

Reviewer: ```suggestion
struct SemanticVersion: Hashable {
```

The @lorentey conjecture: Anything that's substitutable under `Equatable` and is easily capable of being `Hashable` should be.

  var major, minor, patch: Int
}

let semverRegex = Regex {
  TryCapture(OneOrMore(.digit)) { Int($0) }
  "."
  TryCapture(OneOrMore(.digit)) { Int($0) }
  Optionally {
    "."
    TryCapture(OneOrMore(.digit)) { Int($0) }
  }
}.map { _, c1, c2, c3 in
  SemanticVersion(major: c1, minor: c2, patch: c3 ?? 0)

> }

Reviewer: This is such a compelling use

  Capture(b)
}

if let result = try regex.firstMatch(in: "abcdefabcdef") {
  print(result[a]) // => Optional("abc")
  print(result[b]) // => Optional("def")
  print(result[c]) // => nil
}
```

> A regex is considered invalid when it contains a use of reference without it ever being used as the `as:` argument to an initializer of `Capture` or `TryCapture` in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the argument to a `Regex.Match.subscript(_:)` must have been used as the `as:` argument to an initializer of `Capture` or `TryCapture` in the regex that produced the match.

Reviewer: Might be worth adding a sentence or breaking into two paragraphs. It took a few glances for me to remember that the subscript always returns optional, so it's good to call that out explicitly outside of details

  public init<W, C0, Component: RegexComponent>(
    _ component: Component,
    count: Int,
    _ behavior: RepetitionBehavior? = nil
  )
  where Output == (Substring, C0),
        Component.RegexOutput == (Substring, C0),
        R.Bound == Int
  
  public init<W, C0, Component: RegexComponent>(
    count: Int,
    _ behavior: RepetitionBehavior? = nil,
    @RegexComponentBuilder _ component: () -> Component
  )
  where Output == (Substring, C0),
        Component.RegexOutput == (Substring, C0),
        R.Bound == Int
  
  public init<W, C0, Component: RegexComponent, RE: RangeExpression>(
    _ component: Component,
    _ expression: RE,
    _ behavior: RepetitionBehavior? = nil
  )
  where Output == (Substring, C0?),
        Component.RegexOutput == (W, C0),
        R.Bound == Int
  
  public init<W, C0, Component: RegexComponent, RE: RangeExpression>(
    _ expression: RE,
    _ behavior: RepetitionBehavior? = nil,
    @RegexComponentBuilder _ component: () -> Component
  )
  where Output == (Substring, C0?),
        Component.RegexOutput == (W, C0),
        R.Bound == Int
  
  // ... `O(arity)` overloads
}
```

</details>

#### Repetition behavior

> Repetition behavior defines how eager a repetition component should match the input. All repetition regex components have eager behavior by default.

Reviewer: ```suggestion
Repetition behavior defines how eager a repetition component should match the input. Behavior can be unspecified, in which case it will default to `eager` unless an option is provided to change the default (see [Unicode for String Processing](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/Evolution/ProposalOverview.md#unicode-for-string-processing))
```

  /// algorithm by default. To specify a different word boundary algorithm,
  /// see the `RegexComponent.wordBoundaryKind(_:)` method.
  ///
  /// This anchor is equivalent to `\b` in regex syntax.
  public static var wordBoundary: Anchor { get }

  /// An anchor that matches at the start of the input string.
  ///
  /// This anchor is equivalent to `\A` in regex syntax.
  public static var startOfSubject: Anchor { get }

  /// An anchor that matches at the end of the input string.
  ///
  /// This anchor is equivalent to `\z` in regex syntax.
  public static var endOfSubject: Anchor { get }

  /// An anchor that matches at the end of the input string or at the end of
  /// the line immediately before the the end of the string.
  ///
  /// This anchor is equivalent to `\Z` in regex syntax.
  public static var endOfSubjectBeforeNewline: Anchor { get }

  /// An anchor that matches at a grapheme cluster boundary.
  ///
  /// This anchor is equivalent to `\y` in regex syntax.
  public static var textSegmentBoundary: Anchor { get }

  /// An anchor that matches at the first position of a match in the input
  /// string.
  ///

>   /// This anchor is equivalent to `\y` in regex syntax.

Reviewer: ```suggestion
  /// This anchor is equivalent to `\G` in regex syntax.
```

IIRC

Author: `\y` is correct according to @natecook1000

  public init<W, C0, Component: RegexComponent>(
    @RegexComponentBuilder _ component: () -> Component
  ) where Output == (Substring, C0), Component.RegexOutput == (W, C0)
  
  public init<W, C0, C1, Component: RegexComponent>(
    @RegexComponentBuilder _ component: () -> Component
  ) where Output == (Substring, C0, C1), Component.RegexOutput == (W, C0, C1)
  
  // ... `O(arity)` overloads
}
```

</details>

### Composability

Let's put everything together now and parse this example bank statement.

```
CREDIT    04062020    PayPal transfer    $4.99
CREDIT    04032020    Payroll            $69.73
DEBIT     04022020    ACH transfer       $38.25
DEBIT     03242020    IRS tax payment    $52249.98
```

Here we have 2 types of transaction kinds, CREDIT and DEBIT, we have a date
denoted by mmddyyyy, a description, and the amount paid.

```swift
enum TransactionKind: String {
  case credit = "CREDIT"
  case debit = "DEBIT"
}

struct Date {
  var month: Int
  var day: Int
  var year: Int

  init?(mmddyyyy: String) {
    ...
  }
}

let statementRegex = Regex {
  // First, lets capture the transaction kind by wrapping our ChoiceOf in a

>   // TryCapture because we want 

Reviewer 4: Oops, this is my fault I forgot to finish the sentence in my document. Something like, "First, lets capture the transaction kind by wrapping our ChoiceOf in a TryCapture because our initializer can return nil on failure."

  TryCapture {
    ChoiceOf {
      "CREDIT"
      "DEBIT"
    }
  } transform: {
    TransactionKind(rawValue: String($0))
  }

  OneOrMore(.whitespace)

  // Next, lets represent our date as 3 separate repeat quantifiers. The first
  // two will require 2 digit characters, and the last will require 4. Then
  // we'll take the entire substring and try to parse a date out.
  TryCapture {
    Repeat(.digit, count: 2)
    Repeat(.digit, count: 2)
    Repeat(.digit, count: 4)
  } transform: {
    Date(mmddyyyy: String($0))
  }

  OneOrMore(.whitespace)

  // Next, grab the description which can be any combination of word characters,
  // digits, etc.
  Capture {

>     OneOrMore(.any)

Reviewer 3: I added this whole example as a test in https://github.com/apple/swift-experimental-string-processing/pull/344.
